### PR TITLE
Block helper rewrite

### LINKED ISSUE #246: [object Object] for {glossary} when making anki cards through yomitan
**Description**
```html
<div style="text-align: left;">
	<ol>
		<li><i>(1, v5m, vt, JMdict Extra)</i> to read</li>
		<li><i>(1, v5m, vt, JMdict)</i> to read</li>
		<li><i>(2, JMdict Extra)</i> to recite (e.g. a sutra) | to chant</li>
		<li><i>(2, JMdict)</i> to recite (e.g. a sutra) | to chant</li>
		<li><i>(3, JMdict Extra)</i> [object Object]</li>
		<li><i>(3, JMdict)</i> [object Object]</li>
		<li><i>(4, JMdict Extra)</i> to decipher</li>
		<li><i>(4, JMdict)</i> to decipher</li>
		<li><i>(5, JMdict Extra)</i> [object Object]</li>
		<li><i>(5, JMdict)</i> [object Object]</li>
		<li><i>(6, JMdict Extra)</i> [object Object]</li>
		<li><i>(6, JMdict)</i> [object Object]</li>
	</ol>
</div>
```
This is {glossary} for 読め (beautified)

**Browser version**
Firefox nightly 119.0a1 (2023-09-20)
I also tried Firefox stable 117.0.1 (64-bit)

**Yomichan version**
23.8.19.0

**Exported settings file**
[yomitan-settings-2023-09-23-06-58-13.zip](https://github.com/themoeway/yomitan/files/12705830/yomitan-settings-2023-09-23-06-58-13.zip)

### LINKED ISSUE #277: anki: furigana-plain no longer works
**Description**
Using `{furigana-plain}` in a card format field results in `undefined` instead of the standard `日本語[にほんご]` furigana format as described in the help page. This is a regression with respect to the last released version of the original Yomichan.

**Browser version**
Firefox 118.0.1

**Yomichan version**
Yomitan (development build) 23.10.1.0

### PR DESCRIPTION
**Pull request by praschke**
Closes #246, closes #277.

in handlebars, helpers can be passed information through positional or labeled arguments as well as two templated blocks, the truthy `options.fn()` and the falsy `options.else()`. these block arguments can be evaluated under whatever context as many times as the helper wants, and they nominally return strings. block helpers (as opposed to regular helpers) are only useful for control flow or composition of effects.

for some reason, handlebars does not always return strings as documented and specced, but objects. when evaluated, `{{{.}}}` gives the context object, `{{{definition}}}` gives the definition object from the context. in contrast, kibana's no-eval handlebars (written in typescript) returns the stringified counterparts, `'[object Object]'`. when yomichan's helpers attempt to treat the result of blocks as a non-string, things break.

the helpers affected by this behavior are
- `dumpObject`
- `furigana`
- `furiganaPlain`
- `formatGlossary`

uses of these helpers must be rewritten into non-block helpers, as normal arguments are capable of passing javascript objects from the context to the helper.

```handlebars
{{#dumpObject}}{{{.}}}{{/dumpObject}}
{{dumpObject .}}

{{#furigana}}{{{definition}}}{{/furigana}}
{{furigana definition}}

{{#furiganaPlain}}{{{definition}}}{{/furiganaPlain}}
{{furiganaPlain definition}}

{{#formatGlossary ../dictionary}}{{{.}}}{{/formatGlossary}}
{{formatGlossary ../dictionary .}}
```

in general, block helpers have been overused in yomichan, meaning the default templates and templates written by the community have been more verbose than necessary. this PR also changes the default templates and helper documentation to reflect better practices.

another bug was found due to `regexReplace` and `regexMatch` not excluding the final options object from the arguments iterated over, resulting in `[object Object]` sometimes being placed in the output.

the deprecated `kanjiLinks` and `sanitizeCssClass` have been removed.

the `typeof` helper now has non-portable behavior with regard to block arguments. the no-eval path should probably be taken on both chrome and firefox instead of using `compileFnName` to dynamically dispatch.

### CODE REVIEW

**Comment by djahandarie** [comment]
Nice, this looks great!! The more minimal syntax is a nice unexpected improvement.

What is the expected experience for users after this PR? Will that templates upgrade file fix people's templates automatically in all cases, or is some manual action required?

**Comment by praschke** [comment]
JP Mining Note uses `formatGlossary`, and the lines containing them match the lines in the default template. depending on how the template patcher works, it may automatically convert some customized templates. i've only looked at JPMN, there are more customized templates in the wild.

automatic patching also can't be counted on for new users, because someone with an up-to-date install copying JPMN would not be subject to the patch process.

manual work is required to be safe, which is why i included examples of the old and new syntax.

**Comment by djahandarie** [comment]
@praschke Hmm. Could we maybe add a button to "Fix broken yomichan templates" in the settings which would run the template patcher for people? I think the template is idempotent so it should be a fairly safe thing to include.

**Comment by praschke** [comment]
@djahandarie the settings upgrade process happens automatically, and a button would give the impression that all instances would be fixed automatically, when only the default templates are guaranteed to be fixed by the patch. i can't say that all templates in the wild would be fixed by such a button.

**Comment by djahandarie** [comment]
Hmm, "Attempt to fix template" button then? 😄

Or if you think even that would be confusing / overcommitting, maybe an addition to the README would be most appropriate then? I think we need some sort of easily-visible guidance for people regarding how to fix their templates 🤔

**Comment by praschke** [comment]
as well as the readme, i think the welcome page could be used to alert people after an update in case they've modified their templates from the default. i can look into that today or tomorrow.

**Comment by praschke** [comment]
there is now a migration guide in the readme. for new users of yomitan or users migrating from yomichan or yomibaba, the welcome page notes the situation and links to the guide in the readme.

if someone has custom templates and yomitan updates, then the welcome page will open once and a cautionary notification will be highlighted in red with a similar note and link.

File: ext/js/data/options-util.js [hunk 1]
```diff
@@ -997,4 +998,36 @@ class OptionsUtil {
...
+        await this._applyAnkiFieldTemplatesPatch(options, '/data/templates/anki-field-templates-upgrade-v21.handlebars');
+
+        let customTemplates = false;
+        for (const {options: profileOptions} of options.profiles) {
+            if (profileOptions.anki.fieldTemplates !== null) {
+                customTemplates = true;
+            }
+        }
+
+        if (customTemplates && isObject(chrome.storage)) {
```
Comment (djahandarie) [inline ext/js/data/options-util.js:1012]: Forgotten `chrome.storage` check? (I take it you are trying to eliminate these now that all supported browsers are supposed to have `chrome.storage`)

**Review by djahandarie — COMMENTED** [review]
Looks great, I think this should cover most cases for users! Left one minor comment, otherwise good to approve.

File: ext/js/data/options-util.js (hunk 1 above)
Comment (praschke) [reply ext/js/data/options-util.js:1012]: i added the check here to not fail the test harness.

**Review by djahandarie — APPROVED** [review]

**Pull request merged**